[ONNX] Create stand alone diagnostic rule on nearest match finding in dispatcher

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104267

Change the diagnostic call in nearest match finding from UnsupportedNodeAnalysis to its own guarding rule.

titaiwangms committed:
[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Use no_symbolic_function_for_call_function rule on dispatch function in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Create stand alone diagnostic rule on nearest match finding in dispatcher"


Change the diagnostic call in nearest match finding from UnsupportedNodeAnalysis to its own guarding rule.

[ghstack-poisoned]

thiagocrepaldi reviewed (approved):
LGTM, but @BowenBao is the best reference for Diagnostics

titaiwangms commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -354,6 +354,35 @@
+- id: FXE0014
+  name: nearest-match-symbolic-function
+  short_description:
+    text: Can not find the perfect match symbolic function with op schema, but find the nearest match one.
```
Make sense.

titaiwangms replied:
Rename it to `find-opschema-matched-symbolic-function`, and also covers the scenario when perfect match is found (LEVEL=NOTE). Added instructions on WARNING cases.

titaiwangms commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -146,22 +175,23 @@ def _find_the_perfect_or_nearest_match_onnxfunction(
-            f"which should be registered under {node.target}. But a nearest match is found.",
-            unsupported_fx_node=node,
+
+        diagnostic = diagnostic_context.inflight_diagnostic()
```
Yes, it then becomes living under onnxfunction_dispatcher.py

titaiwangms committed:
Update on "[ONNX] Create stand alone diagnostic rule on nearest match finding in dispatcher"


Change the diagnostic call in nearest match finding from UnsupportedNodeAnalysis to its own guarding rule.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -410,8 +410,8 @@ def run_node(
 
-    @diagnostics.diagnose_call(diagnostics.rules.atenlib_fx_to_onnx)
     @_beartype.beartype
+    @diagnostics.diagnose_call(diagnostics.rules.atenlib_fx_to_onnx)
```
Bug fixed. diagnose_call shoud be on top of the target function.

titaiwangms commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -354,6 +354,35 @@
+    text: "Can not find the perfect match symbolic function with op schema, but find the nearest match one."
+    markdown: |
+      The warning message indicates that the dispatcher was unable to find an exact match for the symbolic
+      function based on the provided operation schema and input data types. Instead, it found the closest
```
LMK if you still see such things in the new md.

BowenBao commented on torch/onnx/_internal/diagnostics/rules.yaml:
```diff
@@ -354,6 +354,42 @@
+
+      There are two types of level that can be triggered in this rule:
+
+      1. NOTE: A perfect match is found, and no warning is issued.
```
```suggestion
      1. NONE: A perfect match is found, and no warning is issued.
```
Should be `NONE`. I hate sarif level that NOTE actually also means problem, just not as severe as warning.

BowenBao replied:
Actually don't worry about it here. I found out Sarif extension render it as `Level: note, Kind informational`, although the json file lists it as `Level: none, Kind informational`. We should just drop NONE level from our diagnostics API. And point NOTE to sarif None ...... 💀

titaiwangms replied:
So keep it for now, and there will be a refactor for NOTE/NONE?

BowenBao replied:
yep 🚢

titaiwangms commented:
@pytorchbot merge
